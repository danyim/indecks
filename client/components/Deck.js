import React from 'react';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/Deck';
import CSSTransitionGroup from 'react-addons-css-transition-group';

const Deck = React.createClass({
  navigateToDeck(deckId) {
    browserHistory.push(`/view/${deckId}`);
  },

  render() {
    const { deck, i } = this.props;
    const post = {};

    return (
      <figure className="grid-figure">
        <div className="">
          <h1 className={`${styles['deck-title']}`} onClick={() => this.navigateToDeck(deck.id)}>{deck.title}</h1>
        </div>

        <figcaption>
          {/*
          <p><code>{deck.id}</code></p>
          <p>URL: <a href={deck.url}>Link</a></p>
          */}
          <p className={`${styles['card-count']}`}>{deck.cards.length} cards</p>
          <p>{deck.description}</p>
          <div className="control-buttons">
            <Link className="button" to={`/view/${deck.id}/1`}>
              <span className="comment-count">
                View
              </span>
            </Link>
            <Link className="button" to={`/edit/${deck.id}/1`}>
              <span className="comment-count">
                Edit
              </span>
            </Link>
            <a className="button button-delete" onClick={() => this.props.removeDeck(deck.id)}>
                Delete
            </a>
          </div>
        </figcaption>
      </figure>
    );
  }
})

export default Deck;
