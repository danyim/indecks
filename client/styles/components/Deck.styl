.deck-title
  font-size 3.5rem
  padding 2rem 2rem 1rem 2rem
  cursor pointer

.grid-figure
  display flex
  flex-direction column

.grid-figure > div:first-child
  flex 1 0 auto

.grid-figure figcaption
  padding 0rem 2rem 1rem 2rem

.control-buttons
  padding 0rem 2rem 2rem 2rem
