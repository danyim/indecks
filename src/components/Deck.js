import React from 'react';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/Deck';
import CSSTransitionGroup from 'react-addons-css-transition-group';

class Deck extends React.Component {
  handleRemoveDeck(deckId, removeFn) {
    if(confirm('Are you sure you want to delete this deck?')) {
      removeFn(deckId);
    }
  }

  navigateToDeck(deckId) {
    browserHistory.push(`/view/${deckId}`);
  }

  render() {
    const { deck, i } = this.props;
    const post = {};

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <div>
          <h1 className={`${styles['deck-title']}`} onClick={() => this.navigateToDeck(deck.id)}>{deck.title}</h1>
        </div>

        <figcaption>
          {/*
          <p><code>{deck.id}</code></p>
          <p>URL: <a href={deck.url}>Link</a></p>
          */}
          <p>{deck.cards.length} cards</p>
          <p>{deck.description}</p>
        </figcaption>

        <div className={`control-buttons ${styles['control-buttons']}`}>
          <Link className="button" to={`/view/${deck.id}/1`}>
            <span className="comment-count">
              View
            </span>
          </Link>
          {/*<Link className="button" to={`/edit/${deck.id}/1`}>
            <span className="comment-count">
              Edit
            </span>
          </Link>*/}
          <a className="button btn-delete" onClick={() => this.handleRemoveDeck(deck.id, this.props.removeDeck)}>
              Delete
          </a>
        </div>
      </figure>
    );
  }
}

export default Deck;
