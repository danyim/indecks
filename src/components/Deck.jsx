import React from 'react';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/Deck';

const propTypes = {
  deck: React.PropTypes.object.isRequired
};

const defaultProps = {};

class Deck extends React.Component {
  navigateToDeck(deckId) {
    browserHistory.push(`/view/${deckId}`);
  }

  render() {
    const { deck } = this.props;
    const { id: deckId } = deck;

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <div>
          <h1
            className={`${styles['deck-title']}`}
          >
            <a onClick={() => this.navigateToDeck(deckId)}>
              {deck.title}
            </a>
          </h1>
        </div>

        <figcaption>
          {/*
          <p><code>{deck.id}</code></p>
          <p>URL: <a href={deck.url}>Link</a></p>
          */}
          <p>{deck.cards.length} cards</p>
          <p>{deck.description}</p>
        </figcaption>

        <div className={`control-buttons ${styles['control-buttons']}`}>
          <Link className="button" to={`/view/${deckId}/1`}>
            Play Deck
          </Link>
          {/*
          <a className="button btn-delete" onClick={() => this.props.handleRemoveDeck(deckId)}>
              Delete
          </a>
          */}
        </div>
      </figure>
    );
  }
}

Deck.propTypes = propTypes;
Deck.defaultProps = defaultProps;

export default Deck;
