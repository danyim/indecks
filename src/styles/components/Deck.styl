.deck-title
  font-size: 3.5rem
  padding: 2rem 2rem 1rem 2rem
  cursor: pointer

// figure is added here to give precedence over the base grid-figure
figure.grid-figure
  display: flex
  flex-direction: column

.grid-figure > div:first-child
  flex: 1 0 auto

.grid-figure figcaption
  padding: 0 2rem 1rem 2rem

.control-buttons
  padding: 0 2rem 2rem 2rem
