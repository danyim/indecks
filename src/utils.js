export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

