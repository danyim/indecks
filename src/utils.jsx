import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
