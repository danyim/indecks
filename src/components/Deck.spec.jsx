import toJson from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import Deck from './Deck'

const defaultProps = {
  deck: {
    id: 'ABCDEFG',
    title: 'Deck title',
    description: 'Deck description',
    cards: []
  },
  handleOnClick: () => {},
  style: null,
  children: null
}

describe('Deck', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = shallow(Deck, defaultProps)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should invoke the handleOnClick when the title is clicked', () => {
    const handler = jest.fn()
    const { wrapper } = setup(Deck, {
      ...defaultProps,
      handleOnClick: handler
    })

    const title = wrapper.find('figure.grid-figure h1')
    expect(title.exists()).toBe(true)

    title.simulate('click')
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should apply the correct styles if provided', () => {
    const testStyle = { color: 'red' }
    const { wrapper } = setup(Deck, {
      ...defaultProps,
      style: testStyle
    })

    const figure = wrapper.find('figure')
    expect(figure.get(0).props.style).toBe(testStyle)
  })
})
