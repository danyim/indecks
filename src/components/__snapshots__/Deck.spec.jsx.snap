// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck should render self and subcomponents 1`] = `undefined`;
